### PR TITLE
【GPT-OSS】update sliding_attention layer use flashmask

### PR DESCRIPTION
**Pull request by xiaoguoguo626807**
#### Before submitting

- [ ] Lint code. If there are lint issues, please format the code first.

```shell
# Install and register `pre-commit` in the project folder
pip install pre-commit && pre-commit install

# Process previous code files separately
pre-commit run --file XXXX.py
```

- [ ] Add test cases into `tests` folder. If there are codecov issues, please add tests cases first.

### PR types

New features
### PR changes

Models
### Description

pcard-67164

1. gpt-oss 中使用sink的attention, 改造eager_attention 支持sink 参数
2. _prepare_decoder_attention_mask 支持sliding_window 处理attention_mask
3. gpt-oss 支持pp 并行，tp4+pp2 收敛曲线 vs tp=8 收敛曲线 （layer = 12) 曲线不完全重合是tp不同导致
image

5. gpt-oss 支持flash mask attention， 修复sink_attention 相关问题； 适配sliding_attention， eager+sliding vs flashmask+sliding
image
6. pp-model 支持自定义embedding
7. 暂时使用fake_token 给未选中token 的expert 使用，排除等待无token expert 返回结果导致hang 的问题。ep 并行接入后此处还会修改。

### CODE REVIEW

File: examples/run_finetune.py [hunk 1]
```diff
@@ -45,6 +45,8 @@
     Ernie4_5_MoeForCausalLMPipe,
     Ernie4_5ForCausalLM,
     Ernie4_5ForCausalLMPipe,
+    GptOssForCausalLM,
```
Comment (lugimzzz) [inline examples/run_finetune.py]: 处理一下冲突，已经去掉了这块

File: paddleformers/nn/attention/sink_impl.py [hunk 2]
```diff
@@ -477,7 +477,16 @@ def backward(ctx, grad_output):
             )
             x = (g_ell.unsqueeze(-1) * query).to(query.dtype)
             _, grad_k_extra_repeated, _ = _flash_attention_backward_dispatch(
-                x, query, key_states, key_states, mu_k, lse_k, dropout, causal, scale
+                x,
```
Comment (lugimzzz) [inline paddleformers/nn/attention/sink_impl.py:480]: 这块修改guoxia确认过不？

File: paddleformers/transformers/gpt_oss/modeling.py [hunk 3]
```diff
@@ -125,21 +125,30 @@ def forward(self, hidden_states: paddle.Tensor, router_indices=None, routing_wei
...
                 # we sum on the top_k and on the sequence lenght to get which experts
                 # are hit this time around
-                expert_hitted = paddle.nonzero(
-                    paddle.greater_than(expert_mask.sum(axis=(-1, -2)), paddle.to_tensor(0, dtype=expert_mask.dtype))
-                )
-            for expert_idx in expert_hitted[:]:
+                # expert_hitted = paddle.nonzero(
+                #     paddle.greater_than(expert_mask.sum(axis=(-1, -2)), paddle.to_tensor(0, dtype=expert_mask.dtype))
+                # )
+            for expert_idx in range(self.num_experts):
```
Comment (lugimzzz) [inline paddleformers/transformers/gpt_oss/modeling.py]: 这段修改的原因？

File: paddleformers/transformers/gpt_oss/modeling.py [hunk 4]
```diff
@@ -930,8 +815,16 @@ def forward(
             inputs_embeds = ScatterOp.apply(inputs_embeds)
 
         # embed positions
-        if attn_mask_start_row_indices is not None or get_use_casual_mask():
+        if attn_mask_startend_row_indices is not None or get_use_casual_mask():
```
Comment (lugimzzz) [inline paddleformers/transformers/gpt_oss/modeling.py]: 去掉 or get_use_casual_mask()

File: paddleformers/transformers/gpt_oss/modeling.py [hunk 5]
```diff
@@ -1229,6 +1121,7 @@ def forward(
         attention_mask: Optional[paddle.Tensor] = None,
         inputs_embeds: Optional[paddle.Tensor] = None,
         labels: Optional[paddle.Tensor] = None,
+        loss_mask: Optional[paddle.Tensor] = None,
```
Comment (lugimzzz) [inline paddleformers/transformers/gpt_oss/modeling.py:1117]: 这个一定要吗

File: paddleformers/transformers/gpt_oss/modeling.py (hunk 3 above)
Comment (xiaoguoguo626807) [reply paddleformers/transformers/gpt_oss/modeling.py]: 修 pp 做的尝试，暂时没用可以删掉

File: paddleformers/transformers/gpt_oss/modeling.py (hunk 5 above)
Comment (xiaoguoguo626807) [reply paddleformers/transformers/gpt_oss/modeling.py:1117]: 这个和基础设施对齐，如果删的话一起删不会出问题，就不在这里单独删了

File: paddleformers/transformers/gpt_oss/modeling.py (hunk 4 above)
Comment (xiaoguoguo626807) [reply paddleformers/transformers/gpt_oss/modeling.py]: done

File: paddleformers/nn/attention/sink_impl.py (hunk 2 above)
Comment (xiaoguoguo626807) [reply paddleformers/nn/attention/sink_impl.py:480]: 这里不改会报错，缺参数

**Review by lugimzzz — APPROVED** [review]
LGTM
